token/services/tokens/tokens.go

```diff
@@ -225,6 +233,96 @@ func (t *Tokens) UnsupportedTokensIteratorBy(ctx context.Context, walletID strin
 	return t.Storage.tokenDB.UnsupportedTokensIteratorBy(ctx, walletID, typ)
 }
 
+// PruneInvalidUnspentTokens checks that each unspent token is actually available on the ledger.
+// Those that are not available are deleted.
+// The function returns the list of deleted token ids
+func (t *Tokens) PruneInvalidUnspentTokens(ctx context.Context) ([]*token2.ID, error) {
+	tmsID := t.Storage.tmsID
+	tms, err := t.TMSProvider.GetManagementService(token.WithTMSID(tmsID))
+	if err != nil {
+		return nil, errors.WithMessagef(err, "failed getting token management service [%s]", tmsID)
+	}
+	// network
+	tmsID = tms.ID()
+	net, err := t.NetworkProvider.GetNetwork(tmsID.Network, tms.Channel())
+	if err != nil {
+		return nil, errors.WithMessagef(err, "failed getting network [%s:%s]", tmsID.Network, tmsID.Channel)
+	}
+
+	// get unspent tokens
+	it, err := tms.Vault().NewQueryEngine().UnspentTokensIterator()
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to get an iterator of unspent tokens")
+	}
+	defer it.Close()
+
+	var deleted []*token2.ID
+	var buffer []*token2.UnspentToken
+	bufferSize := 50
```

Comment:
This buffer size doesn't look very correct. I guess your limitation is the chaincode? We should pass all tokens and then the chain code should split them into batches and return the response. 3 reasons:
* If we have a different implementation that supports bigger chunks, we cannot leverage it.
* All subsequent calls (e.g. SpentIds, deleteTokens) cannot delete faster.
* It is not clear what the reason, i.e. limitation is. Is it the cc?
The APIs should be agnostic to the limitations of the implementation.

Reply:
will open a Github issue

Reply:
https://github.com/hyperledger-labs/fabric-token-sdk/issues/988